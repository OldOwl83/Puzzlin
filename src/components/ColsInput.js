import { useContext, useState, useEffect } from "react";
import { PuzzleGrid } from "../RompeCocosApp";

export const ColsInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ input, setInput ] = useState( puzzleGrid[0] );
    
    useEffect(() => {
        
        if( input >= 3 || input == '' )
        {
            if( input >= 3 )
                setPuzzleGrid( [ Number( input ), puzzleGrid[1] ] );
            else
                setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
        }else
        {
            setPuzzleGrid( [ Number( 3 ), puzzleGrid[1] ] );
            setInput( 3 );
        }
    }, [ input, setPuzzleGrid ]);

    return <input 
                type="number" 
                value={ input }
                min="3"
                onChange={ ( e ) => setInput( e.target.value) } 
            />
}
