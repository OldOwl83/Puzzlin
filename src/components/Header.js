import React from 'react';
import { Link, useLinkClickHandler } from 'react-router-dom';

export const Header = () => {

    const linker = useLinkClickHandler( '/', { 
        target: '_self',
        replace: false,
        state: {
            contScroll: 0,
            selectedPhoto: 'images/portada1.png',
            author: 'Mauro Donnantuoni',
        }
    });

    return (
        <>
        <Link to='/' id='toHomeLink' onClick={ linker }>
            <header>
                <h1>Puzzlin</h1>
                <div id="logoContainer">
                    <img src="images/rompecabezas.png" alt="logo" />
                </div>
            </header>
        </Link>

        <hr />
        </>
    );
};
