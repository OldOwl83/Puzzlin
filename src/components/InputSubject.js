import { useState } from "react";
import { PropTypes } from 'prop-types';
import { useNavigate } from "react-router-dom";

import '../styles/InputSubject.css';

export const InputSubject = ( { setSearch } ) => {

    // console.log( 'InputSubject' );

    const navigate = useNavigate();

    const [ input, setInput ] = useState( '' );
    
    const submitInput = (e) => {
        e.preventDefault();
        
        setInput( '' );

        setSearch( input );

        navigate( `../${ input }` );
    }

    return (

        <h2>Input a subject and select a photo. Then, press on the preview to start.
            <form onSubmit={ submitInput }>
                <input 
                    id="inputSubject"
                    type="text"
                    autoComplete="on"
                    value={ input }
                    onChange={ (e) => setInput( e.target.value ) }
                />
            </form>
        </h2>
    )
}

InputSubject.propTypes = {

    setSearch: PropTypes.func.isRequired,
}