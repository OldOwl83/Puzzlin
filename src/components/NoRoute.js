import React from 'react';

export const NoRoute = () => {
  return <p className='message'>Nonexistent path.</p>;
};
