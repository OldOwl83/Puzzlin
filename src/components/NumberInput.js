/*
* General purpose number input. It receives by parameters a function which changes an external value or state, the minimum and maximum numbers allowed and the initial value. If the user tries to exceed the set range, it automatically sets the minimum or maximum accordingly. It sets the value as an integer data type.
*/

import { useState, useEffect, memo } from "react";

export const NumberInput = memo(( { setValue, min, max, initialValue }) => {

    const [ input, setInput ] = useState( initialValue );
    
    useEffect(() => {
    
        if( ( input >= min && input <= max ) || input === '' ) //It must include '' to allow erase manually the text in the input box.
        {
            if( input >= min && input <= max )
                setValue( Number( input ) );
            else 
                setValue( Number( min ) ); // If the input box remain empty, the set value is the minimum.
        }else if( input > max )
        {
            setValue( Number( max ) );
            setInput( max );
        }else
        {
            setValue( Number( min ) )
            setInput( min );
        }
    }, [ input, min, max, setValue ]);

    return (
    
        <input 
            type="number" 
            value={ input }
            min={ min }
            max={ max }
            onChange={ ( e ) => setInput( e.target.value ) } 
        />
    );
});
