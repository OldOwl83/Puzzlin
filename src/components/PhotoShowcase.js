/*
*
*/

import { useState, useEffect } from "react";
import { useLocation, useNavigate } from "react-router-dom";

import '../styles/PhotoShowcase.css';

export const PhotoShowcase = ( { photos = [] } ) => {
    
    // console.log(photos);

    const location = useLocation();
    const navigate = useNavigate();

    const [ photosContScroll, setPhotosContScroll ] = useState( 0 );
    const [ photosContHeight, setPhotosContHeight ] = useState( 0 );

    useEffect( () => {
        setPhotosContHeight( photos.length * 110 );
        setPhotosContScroll( location.state?.contSroll || 0 );

    }, [ photos, location ]);

    return (
        <div id="sampler">

        {/* Scroll up button */}
            <div
                className="material-icons arrows"
                onClick={ () => {
                    ( photosContScroll + 330 <= 0 ) ? setPhotosContScroll( photosContScroll + 330 ) : setPhotosContScroll( 0 );
                } }    
            >keyboard_double_arrow_up</div>

        {/* Photos Showcase */}
            <div id="photosHidder">
                <div 
                    id="photosContainer" 
                    style={ { top: photosContScroll, height: photosContHeight } }
                >
                    
                    { photos.map( ( { id, description, urlSmall, urlRegular, author } ) => {
                    
                            return (

                                <div key={ id } className="photos">

                                    <img 
                                        className="animate__animated animate__zoomIn" 
                                        src={ urlSmall } 
                                        alt={ description }
                                        draggable='false' 
                                        onClick={ () => {

                                            navigate( location, {
                                                replace: true,
                                                state: {
                                                    contSroll: photosContScroll,
                                                    selectedPhoto: urlRegular,
                                                    author: author,
                                                }
                                            })
                                        }}
                                    />

                                </div>)
                    })}
                    
                </div>
            </div>

        {/* Scroll down button */}
            <div 
                className="material-icons arrows"
                onClick={ () => {
                    ( photosContScroll - 330 >= 440 - photosContHeight ) ? setPhotosContScroll( photosContScroll - 330 ) : setPhotosContScroll( 440 - photosContHeight );
                }}    
            >keyboard_double_arrow_down</div>
        </div>
    )
}
