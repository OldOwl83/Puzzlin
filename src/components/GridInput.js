/*
* It handles the contextual values of the puzzle grid. These values represent the columns and rows to cut out the puzzle photo and are in an array with two indices, which is a state belonging to the <Puzzlin /> component.
*/

import { useContext, useState, useEffect } from "react";

import { PuzzleGrid } from "../Puzzlin";
import { NumberInput } from "./NumberInput";

import "../styles/GridInput.css";

export const GridInput = () => {

    const { puzzleGrid, setPuzzleGrid } = useContext( PuzzleGrid );
    
    const [ cols, setCols ] = useState( puzzleGrid[0] );
    const [ rows, setRows ] = useState( puzzleGrid[1] );
    
    useEffect(() => {
        
        setPuzzleGrid( [ cols, rows ] );
        
    }, [ setPuzzleGrid, cols, rows ]);

    return (
    
        <div id="GridInput">
            <p>Columns
                <NumberInput setValue={ setCols } min={ 4 } max={ 8 } initialValue={ puzzleGrid[0] } />
                {/* It send de previous context values to initial value of this component, to maintain them across all routes. */}
            </p>
            <p>Rows
                <NumberInput setValue={ setRows } min={ 3 } max={ 8 } initialValue={ puzzleGrid[1] } />
            </p>
        </div>
    );
}
