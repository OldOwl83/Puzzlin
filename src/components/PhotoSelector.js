import { useState, useEffect, useContext } from "react";

import { InputSubject } from "./InputSubject";
import { Preview } from './Preview';
import { useFetchPhotos } from "../hooks/useFetchPhotos";
import { PhotoShowcase } from "./PhotoShowcase";

import { MainPhotoContext } from "../RompeCocosApp";

import '../styles/PhotoSelector.css';
import 'animate.css';

export const PhotoSelector = () => {

    console.log('PhotoSelector');

    const [search, setSearch] = useState( '' );

    const { setPhotoPuzzle } = useContext( MainPhotoContext );

    // const [ imgSelected, setImgSelected] = useState( "images/tapa.png" );

    const { photos, loading } = useFetchPhotos( search );

    useEffect(() => {
        if(photos.length > 0)
            setPhotoPuzzle( photos[0].urlRegular );
    }, [ photos ])

    return (
        <>
            <InputSubject 
                setSearch={ setSearch } 
             />

            <div id="selectorContainer">
                
                <PhotoShowcase photos={ photos } />

                { loading && <p className="message">Loading...</p> }
                    
                { !photos && <p className="message">Connection to images provider failed.</p> }

                { search !== '' && !loading && photos.length < 1 && <p className="message">There were no matches with the search words.</p> }

                { (search === '' || photos.length > 0) && !loading && <Preview /> }
            </div>
        </>
    )
}