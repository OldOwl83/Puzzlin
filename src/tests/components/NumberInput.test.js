import { shallow, mount } from 'enzyme';

import { NumberInput } from "../../components/NumberInput";

describe('Tests on <NumberInput /> component', () => {
  
    const setValue = jest.fn();

    const min = 3;
    const max = 8;
    const init = 5;

    test('It should match the snapshot and takes the initial value', () => {
        
        const wrapper = shallow( <NumberInput 
                                    setValue={ setValue } 
                                    min={ min } 
                                    max={ max } 
                                    initialValue={ init } 
                                /> );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( init );
    });

    test('It should call "setValue" function with the same value entered', () => {
      
        const wrapper = mount( <NumberInput 
                                    setValue={ setValue } 
                                    min={ min } 
                                    max={ max } 
                                    initialValue={ init } 
                                /> );

        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( init );
        
        wrapper.find( 'input' ).simulate( 'change', { target: { value: 4 }});
        
        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( 4 );

        expect( setValue ).toHaveBeenCalledTimes( 2 );
        expect( setValue ).toHaveBeenCalledWith( init );
        expect( setValue ).toHaveBeenCalledWith( 4 );
    });

    test('It should call "setValue" function with the minimum value', () => {
      
        const wrapper = mount( <NumberInput 
                                    setValue={ setValue } 
                                    min={ min } 
                                    max={ max } 
                                    initialValue={ init } 
                                /> );

        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( init );
        
        wrapper.find( 'input' ).simulate( 'change', { target: { value: 1 }});
        
        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( min );

        expect( setValue ).toHaveBeenCalledTimes( 3 );
        expect( setValue ).toHaveBeenCalledWith( init );
        expect( setValue ).toHaveBeenCalledWith( min );
    });

    test('It should call "setValue" function with the maximum value', () => {
      
        const wrapper = mount( <NumberInput 
                                    setValue={ setValue } 
                                    min={ min } 
                                    max={ max } 
                                    initialValue={ init } 
                                /> );

        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( init );
        
        wrapper.find( 'input' ).simulate( 'change', { target: { value: 10 }});
        
        expect( wrapper.find( 'input' ).prop( 'value' )).toBe( max );

        expect( setValue ).toHaveBeenCalledTimes( 3 );
        expect( setValue ).toHaveBeenCalledWith( init );
        expect( setValue ).toHaveBeenCalledWith( max );
    });
});
