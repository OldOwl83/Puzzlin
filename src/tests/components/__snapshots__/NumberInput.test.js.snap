// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on <NumberInput /> component It should match the snapshot and takes the initial value 1`] = `
<input
  max={8}
  min={3}
  onChange={[Function]}
  type="number"
  value={5}
/>
`;

exports[`Tests on <NumberInput /> component It should match with the snapshot and takes the initial value 1`] = `
<input
  max={8}
  min={3}
  onChange={[Function]}
  type="number"
  value={5}
/>
`;
